pennylane_lightning/lightning_tensor/_tensornet.py

```diff
@@ -31,44 +31,91 @@
 from pennylane.wires import Wires
 
 
-def svd_split(M, bond_dim):
-    """SVD split a matrix into a matrix product state via numpy linalg. Note that this function is to be moved to the C++ layer."""
-    U, S, Vd = np.linalg.svd(M, full_matrices=False)
+def svd_split(Mat, site_shape, max_bond_dim):
+    """SVD decomposition of a matrix via numpy linalg. Note that this function is to be moved to the C++ layer."""
+    U, S, Vd = np.linalg.svd(Mat, full_matrices=False)
     U = U @ np.diag(S)  # Append singular values to U
     bonds = len(S)
-    Vd = Vd.reshape(bonds, 2, -1)
-    U = U.reshape((-1, 2, bonds))
+
+    Vd = Vd.reshape(tuple([bonds] + site_shape + [-1]))
+    U = U.reshape(tuple([-1] + site_shape + [bonds]))
 
     # keep only chi bonds
-    chi = np.min([bonds, bond_dim])
-    U, S, Vd = U[:, :, :chi], S[:chi], Vd[:chi]
+    chi = np.min([bonds, max_bond_dim])
+    U, Vd = U[..., :chi], Vd[:chi]
     return U, Vd
 
 
-def dense_to_mps(psi, n_wires, bond_dim):
-    """Convert a dense state vector to a matrix product state."""
+def decompose_dense(psi, n_wires, site_shape, max_bond_dim):
+    """Decompose a dense state vector/gate matrix into MPS/MPO sites."""
     Ms = [[] for _ in range(n_wires)]
+    site_len = np.prod(site_shape)
+    psi = np.reshape(psi, (site_len, -1))
 
-    psi = np.reshape(psi, (2, -1))  # split psi[2, 2, 2, 2..] = psi[2, (2x2x2...)]
-    U, Vd = svd_split(psi, bond_dim)  # psi[2, (2x2x..)] = U[2, mu] Vd[mu, (2x2x2x..)]
+    U, Vd = svd_split(psi, site_shape, max_bond_dim)
 
-    Ms[0] = U
+    Ms[0] = U.reshape(tuple(site_shape + [-1]))
     bondL = Vd.shape[0]
     psi = Vd
 
     for i in range(1, n_wires - 1):
-        psi = np.reshape(psi, (2 * bondL, -1))  # reshape psi[2 * bondL, (2x2x2...)]
-        U, Vd = svd_split(psi, bond_dim)  # psi[2, (2x2x..)] = U[2, mu] Vd[mu, (2x2x2x..)]
+        psi = np.reshape(psi, (site_len * bondL, -1))
+        U, Vd = svd_split(psi, site_shape, max_bond_dim)
         Ms[i] = U
 
         psi = Vd
         bondL = Vd.shape[0]
 
-    Ms[n_wires - 1] = Vd
+    Ms[-1] = Vd.reshape(tuple([-1] + site_shape))
 
     return Ms
 
 
+def gate_matrix_decompose(gate_ops_matrix, wires, c_dtype):
```

Comment:
This is not producing the right shapes:

```
import scipy
n = 3
hermitian = np.random.rand(2**n, 2**n)
hermitian = hermitian @ hermitian.conj().T

gate = scipy.linalg.expm(1j * hermitian)
# gate = qml.matrix(qml.Toffoli((0, 1, 2)))
mpos, sorted_wires = gate_matrix_decompose(gate, range(3), complex)

[_.shape for _ in mpos]
# [(2, 4, 2), (2, 4, 2, 4), (2, 2, 4)]
```

the last site should be `(4, 2, 2)` so there is a transpose missing at the end

Reply:
Note that from the decomposed MPO I am _not_ getting the original unitary back, so there is a bug somewhere! You can test it with the following code

```python
import scipy
n = 3
hermitian = np.random.rand(2**n, 2**n)
hermitian = hermitian @ hermitian.conj().T

gate = scipy.linalg.expm(1j * hermitian)
original_gate = gate.copy() # for later to double check
# gate = qml.matrix(qml.Toffoli((0, 1, 2)))

mpos, sorted_wires = gate_matrix_decompose(gate, range(3), complex)

print([_.shape for _ in mpos]) # Note that this is not the same order as you describe above
# [(2, 4, 2), (2, 4, 2, 4), (2, 2, 4)]

# recreate unitary
unitary = np.tensordot(mpos[0], mpos[1], axes=([1], [1]))    # s1 [vR] s1' | s2, [vL], s2' vR = s1 s1' s2, s2' vR
unitary = np.tensordot(unitary, mpos[2], axes=([-1], [-1])) # s1 s1' s2, s2' [vR] | s3 s3' [vL] = s1 s1' s2 s2' s3 s3'
unitary = np.transpose(unitary, (0, 3, 1, 4, 2, 5)) # s1 s2 s3 s1' s2' s3'
unitary = np.reshape(unitary, (2**3, 2**3))

np.allclose(unitary, original_gate.conj().T), np.allclose(unitary, original_gate)
# (False, False)

Reply:
Actually, the `gate_matrix_decompose` permute the C-style MPO results of `dense_compose` to the Fortran data layer out to meet the requirement of cutensornet. So it's normal that the above tests would fail. Could you check if the following tests work:

```
n = 3  
hermitian = np.random.rand(2**n, 2**n)  
hermitian = hermitian @ hermitian.conj().T  

gate = scipy.linalg.expm(1j * hermitian)  
original_gate = gate.copy() # for later to double check  
# gate = qml.matrix(qml.Toffoli((0, 1, 2)))  

shape = [2, 2]
mpos = decompose_dense(gate, 3, shape, 128)  

print([_.shape for _ in mpos]) # Note that this is not the same order as you describe above  
# [(2, 4, 2), (2, 4, 2, 4), (2, 2, 4)]  

# recreate unitary  
unitary = np.tensordot(mpos[0], mpos[1], axes=([2], [0]))    # s1 [vR] s1' | s2, [vL], s2' vR = s1 s1' s2, s2' vR  
unitary = np.tensordot(unitary, mpos[2], axes=([-1], [0])) # s1 s1' s2, s2' [vR] | s3 s3' [vL] = s1 s1' s2 s2' s3 s3'  
unitary = np.transpose(unitary, (0, 3, 1, 4, 2, 5)) # s1 s2 s3 s1' s2' s3'  
unitary = np.reshape(unitary, (2**3, 2**3))  

np.allclose(unitary, original_gate.conj().T), np.allclose(unitary, original_gate) 
```

Will update the docstring

Reply:
What shapes should I expect from `decompose_dense` - did you check to reconstruct the unitary matrix and if it is the same as the original unitary? That would be a good sanity check either way for the implementation

Reply:
we can expect [(2, 2, 4), (4, 2, 2, 4), (4, 2, 2)] from ``decompose_dense`` and with the code I posted above, we can reconstruct the original unitary.